book/src/internals/type-system.md

```diff
@@ -0,0 +1,224 @@
+flexibility, all while being conceptually simple.
+
+**Composable** - as with transformation, we'd want types to compose together.
+
+Using Python, JavaScript, C++ or Rust, one could define many different a data
+structure that would correspond to our idea of "relation". Most of them would be
+an object/struct that has column names and types and then a generic array of
+arrays for rows.
+
+PRQL's type system should also be able to express relation as composed from
+primitive types, but have only one idiomatic way of doing so.
+
+In practice this means that builtin types include only primitives (int, text,
+bool, float), struct (for product), enum (for sum) and list (for repeating).
+
+An SQL row would translate to struct, and a relation would translate to a list
+of structs.
+
+I would also strive for the type system to be minimal - don't differentiate
+between tuples, objects and structs. Choose one and stick to it.
+
+**Type constraints** - constrain a type with a predicate. For example, have a
+type of int64 that are equal or greater than 10. Postgres
+[does support this](https://news.ycombinator.com/item?id=34835063). The primary
+value of using constrained types would not be validation (as it is used in
+linked article), but when matching the type.
+
+Say, for example, that we have a pipeline like this:
+
+```
+derive color = switch [x => 'red', true => 'green']
+derive is_red = switch [color == 'red' => true, color == 'green' => false]
+```
+
+It should be possible to infer that `color` is of type `text`, but only when
+equal to `'red'` or `'green'`. This means that the second switch covers all
+possible cases and `is_red` cannot be `null`.
+
+## Physical layout
+
+_Logical type_ is user-facing the notion of a type that is the building block of
+the type system.
+
+_Physical layout_ is the underlying memory layout of the data represented by a
+variable.
+
+In many programming languages, physical layout of a logical type is dependent on
+the target platform. Similarly, physical layout of a PRQL logical type is
+dependent on representation of that type in target STS.
+
+```
+PTS logical type  --->  STS logical type  ---> STS physical layout
+```
+
+Note that STS types do not have a single physical layout. Postgres has a logical
+type `anyelement`, which is a super type of any data type. It can be used a
+function parameter type, but does not have a single physical layout so it cannot
+be used a in column declarations.
+
+For now, PRQL does not define a physical layout of any type. It is not needed
+since PRQL is not used for DDL (see section "Built-in primitives") or does not
+support raw access to underlying memory.
+
+As a consequence, results of a PRQL query cannot be robustly compared across
+DBMSs, since the physical layout of the result will vary.
+
+In the future, PRQL may define a common physical layout of types, probably using
+Apache Arrow.
+
+## Syntax
+
+```
+# built-in types
+type int
```

Comment:
> @vanillajonathan 
> I really like the data types int32 and int64. I've been thinking about database type systems and those two data types are exactly what I had in mind.
>
> But what about int8 and int16?
> We would need something to map to tinyint and smallint.
> 
> Note: Interestingly enough Rust have a [i128](https://doc.rust-lang.org/std/primitive.i128.html) and a [u128](https://doc.rust-lang.org/std/primitive.u128.html) data type.

Do you mean naming `int32, int64` as opposed to `integer, bigint` or `i32 and i64` or `int, long, long long`? That's a clear benefit of defining a new language and defining things to be consistent instead of just adding a `long` before your type name.

But as noted in section "Built-in primitives", we will not be adding these types soon, as there is no need to do so.

Reply:
I would argue you'd need int32, int64, float64, etc. and even though you are only proposing logical PRQL data type primitives here, having them would allow to define operations on large numbers.

Reply:
We can define operations on all kinds of integers by saying: use operator "+" of our target (SQL). Because all SQL dialects support "+" for all int types, this works.

Am I missing something? Is there an operation on ints that has different syntax in SQL depending on the int size?

Reply:
maybe operations was not a good example. What you are missing is that all the modern data stack tools I've seen in the past year or so now use arrow data types and arrow data, and databases are migrating to vectorized data too. I'd like it more if you just adopted them like everyone else, even if what you are defining is abstract data types.

Reply:
If we decide on the proposed abstract types now, we can still adopt Arrow as the standard physical layout and expand the abstract types with primitives that correspond to Arrow logical types.

This was actually my intention, but I don't want to "just adopt" anything for the sake of standardization.

Reply:
Perfect! in that case, I am good with keeping it simple and refine/add more types as needed based on use cases we might encounter when adding other system outputs like query plans from Data Fusion or even generating Polars Py code. I think you got the base types defined well. The rest we can sort out in a way you proposed defining int32.